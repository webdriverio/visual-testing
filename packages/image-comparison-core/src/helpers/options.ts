import {
    DEFAULT_COMPARE_OPTIONS,
    DEFAULT_FORMAT_STRING,
    DEFAULT_SHADOW,
    DEFAULT_TABBABLE_OPTIONS,
    FULL_PAGE_SCROLL_TIMEOUT,
    STORYBOOK_FORMAT_STRING,
} from './constants.js'
import type { ClassOptions, DefaultOptions } from './options.interfaces.js'
import type { MethodImageCompareCompareOptions, ScreenMethodImageCompareCompareOptions } from '../methods/images.interfaces.js'
import type { BeforeScreenshotOptions } from './beforeScreenshot.interfaces.js'
import type { AfterScreenshotOptions } from './afterScreenshot.interfaces.js'
import type { BeforeScreenshotResult } from './beforeScreenshot.interfaces.js'
import type { InstanceData } from '../methods/instanceData.interfaces.js'
import type { ComparisonIgnoreOption } from '../resemble/compare.interfaces.js'
import {
    logAllDeprecatedCompareOptions,
    isStorybook,
    getBooleanOption,
    createConditionalProperty,
    getMethodOrWicOption,
} from './utils.js'

/**
 * Determine the default options by merging user options with sensible defaults
 */
export function defaultOptions(options: ClassOptions): DefaultOptions {
    const isStorybookMode = isStorybook()

    return {
        /**
         * Module-specific options
         */
        addressBarShadowPadding: options.addressBarShadowPadding ?? DEFAULT_SHADOW.ADDRESS_BAR,
        autoElementScroll: getBooleanOption(options, 'autoElementScroll', true),
        addIOSBezelCorners: options.addIOSBezelCorners ?? false,
        autoSaveBaseline: options.autoSaveBaseline ?? true,
        clearFolder: options.clearRuntimeFolder ?? false,
        userBasedFullPageScreenshot: options.userBasedFullPageScreenshot ?? false,
        enableLegacyScreenshotMethod: options.enableLegacyScreenshotMethod ?? false,
        formatImageName: options.formatImageName ?? (isStorybookMode ? STORYBOOK_FORMAT_STRING : DEFAULT_FORMAT_STRING),
        isHybridApp: options.isHybridApp ?? false,
        // Running in storybook mode with multiple browsers can generate many images
        // Defaulting this to true provides better overview for users
        savePerInstance: options.savePerInstance ?? isStorybookMode,
        toolBarShadowPadding: options.toolBarShadowPadding ?? DEFAULT_SHADOW.TOOL_BAR,

        /**
         * Module and method options
         */
        disableBlinkingCursor: options.disableBlinkingCursor ?? false,
        disableCSSAnimation: options.disableCSSAnimation ?? false,
        enableLayoutTesting: options.enableLayoutTesting ?? false,
        fullPageScrollTimeout: options.fullPageScrollTimeout ?? FULL_PAGE_SCROLL_TIMEOUT,
        // Default to false for storybook mode as element screenshots use W3C protocol without scrollbars
        // This also saves an extra webdriver call
        hideScrollBars: getBooleanOption(options, 'hideScrollBars', !isStorybookMode),
        waitForFontsLoaded: options.waitForFontsLoaded ?? true,

        /**
         * Compare options (merged sequentially):
         * 1. Default options (fallback)
         * 2. Root compareOptions (deprecated but supported)
         * 3. User-provided compareOptions
         */
        compareOptions: {
            ...DEFAULT_COMPARE_OPTIONS,
            ...logAllDeprecatedCompareOptions(options),
            ...options.compareOptions,
        },

        /**
         * Tabbable options with deep merging
         */
        tabbableOptions: {
            circle: {
                ...DEFAULT_TABBABLE_OPTIONS.circle,
                ...options.tabbableOptions?.circle,
            },
            line: {
                ...DEFAULT_TABBABLE_OPTIONS.line,
                ...options.tabbableOptions?.line,
            },
        },
    }
}

/**
 * Determine the screen method compare options with proper type safety
 */
export function screenMethodCompareOptions(
    options: ScreenMethodImageCompareCompareOptions,
): ScreenMethodImageCompareCompareOptions {
    return {
        ...createConditionalProperty('blockOutSideBar' in options, 'blockOutSideBar', options.blockOutSideBar),
        ...createConditionalProperty('blockOutStatusBar' in options, 'blockOutStatusBar', options.blockOutStatusBar),
        ...createConditionalProperty('blockOutToolBar' in options, 'blockOutToolBar', options.blockOutToolBar),
        ...methodCompareOptions(options),
    }
}

/**
 * Determine the method compare options with improved type safety
 */
export function methodCompareOptions(options: MethodImageCompareCompareOptions): MethodImageCompareCompareOptions {
    const compareOptionKeys: (keyof MethodImageCompareCompareOptions)[] = [
        'blockOut',
        'ignoreAlpha',
        'ignoreAntialiasing',
        'ignoreColors',
        'ignoreLess',
        'ignoreNothing',
        'rawMisMatchPercentage',
        'returnAllCompareData',
        'saveAboveTolerance',
        'scaleImagesToSameSize',
    ]

    return compareOptionKeys.reduce((result, key) => {
        if (key in options && options[key] !== undefined) {
            result[key] = options[key] as any // Type assertion needed due to union types
        }
        return result
    }, {} as MethodImageCompareCompareOptions)
}

/**
 * Creates BeforeScreenshotOptions by extracting common options from method and wic configurations
 */
export function createBeforeScreenshotOptions(
    instanceData: any,
    methodOptions: {
        hideElements?: HTMLElement[]
        removeElements?: HTMLElement[]
        disableBlinkingCursor?: boolean
        disableCSSAnimation?: boolean
        enableLayoutTesting?: boolean
        hideScrollBars?: boolean
        waitForFontsLoaded?: boolean
    },
    wicOptions: {
        addressBarShadowPadding: number
        toolBarShadowPadding: number
        disableBlinkingCursor?: boolean
        disableCSSAnimation?: boolean
        enableLayoutTesting?: boolean
        hideScrollBars?: boolean
        waitForFontsLoaded?: boolean
    }
): BeforeScreenshotOptions {
    return {
        instanceData,
        addressBarShadowPadding: wicOptions.addressBarShadowPadding,
        disableBlinkingCursor: getMethodOrWicOption(methodOptions, wicOptions, 'disableBlinkingCursor') || false,
        disableCSSAnimation: getMethodOrWicOption(methodOptions, wicOptions, 'disableCSSAnimation') || false,
        enableLayoutTesting: getMethodOrWicOption(methodOptions, wicOptions, 'enableLayoutTesting') || false,
        hideElements: methodOptions.hideElements || [],
        noScrollBars: getMethodOrWicOption(methodOptions, wicOptions, 'hideScrollBars') || false,
        removeElements: methodOptions.removeElements || [],
        toolBarShadowPadding: wicOptions.toolBarShadowPadding,
        waitForFontsLoaded: getMethodOrWicOption(methodOptions, wicOptions, 'waitForFontsLoaded') || false,
    }
}

export interface BuildAfterScreenshotOptionsInput {
    // Required inputs
    base64Image: string
    folders: { actualFolder: string }
    tag: string
    isNativeContext: boolean
    instanceData: InstanceData
    wicOptions: {
        formatImageName: string
        savePerInstance: boolean
    }

    // Optional inputs (for web commands)
    enrichedInstanceData?: BeforeScreenshotResult
    beforeOptions?: BeforeScreenshotOptions
}

/**
 * Builds AfterScreenshotOptions consistently across all commands
 * Handles differences between native and web commands automatically
 */
export function buildAfterScreenshotOptions({
    base64Image,
    folders,
    tag,
    isNativeContext,
    instanceData,
    enrichedInstanceData,
    beforeOptions,
    wicOptions
}: BuildAfterScreenshotOptionsInput): AfterScreenshotOptions {
    // Use enriched data when available (web commands), fallback to instance data (native commands)
    const dataSource = enrichedInstanceData || instanceData

    // Extract common properties with smart fallbacks
    const {
        browserName,
        browserVersion,
        deviceName,
        platformName,
        platformVersion,
    } = dataSource

    // Handle dimension data - enriched data has nested structure, instance data is flat
    const dimensions = enrichedInstanceData?.dimensions?.window
    const devicePixelRatio = dimensions?.devicePixelRatio ?? instanceData.devicePixelRatio
    const isLandscape = dimensions?.isLandscape ?? false
    const outerHeight = dimensions?.outerHeight
    const outerWidth = dimensions?.outerWidth
    const screenHeight = dimensions?.screenHeight ?? instanceData.deviceRectangles?.screenSize?.height
    const screenWidth = dimensions?.screenWidth ?? instanceData.deviceRectangles?.screenSize?.width

    // Handle metadata with smart defaults
    const isMobile = enrichedInstanceData?.isMobile ?? instanceData.isMobile
    const isTestInBrowser = enrichedInstanceData?.isTestInBrowser ?? !isNativeContext
    const logName = enrichedInstanceData?.logName ?? instanceData.logName
    const name = enrichedInstanceData?.name ?? instanceData.name

    const { formatImageName, savePerInstance } = wicOptions

    const afterOptions: AfterScreenshotOptions = {
        actualFolder: folders.actualFolder,
        base64Image,
        filePath: {
            browserName,
            deviceName,
            isMobile,
            savePerInstance,
        },
        fileName: {
            browserName,
            browserVersion,
            deviceName,
            devicePixelRatio: devicePixelRatio || NaN,
            formatImageName,
            isMobile,
            isTestInBrowser,
            logName,
            name,
            outerHeight: outerHeight || NaN,
            outerWidth: outerWidth || NaN,
            platformName,
            platformVersion,
            screenHeight: screenHeight || NaN,
            screenWidth: screenWidth || NaN,
            tag,
        },
        isLandscape,
        isNativeContext,
        platformName: instanceData.platformName,
    }

    // Add browser state options only for web commands (when beforeOptions is provided)
    if (beforeOptions) {
        afterOptions.disableBlinkingCursor = beforeOptions.disableBlinkingCursor
        afterOptions.disableCSSAnimation = beforeOptions.disableCSSAnimation
        afterOptions.enableLayoutTesting = beforeOptions.enableLayoutTesting
        afterOptions.hideElements = beforeOptions.hideElements
        afterOptions.hideScrollBars = beforeOptions.noScrollBars
        afterOptions.removeElements = beforeOptions.removeElements
    }

    return afterOptions
}

/**
 * Prepare ignore options for resemble.js comparison
 */
export function prepareIgnoreOptions(imageCompareOptions: MethodImageCompareCompareOptions): ComparisonIgnoreOption[] {
    const resembleIgnoreDefaults: ComparisonIgnoreOption[] = ['alpha', 'antialiasing', 'colors', 'less', 'nothing']

    return resembleIgnoreDefaults.filter((option) =>
        Object.keys(imageCompareOptions).find(
            (key: keyof typeof imageCompareOptions) => key.toLowerCase().includes(option) && imageCompareOptions[key],
        ),
    )
}

