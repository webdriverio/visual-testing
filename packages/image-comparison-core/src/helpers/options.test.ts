import { describe, it, expect } from 'vitest'
import { defaultOptions, methodCompareOptions, screenMethodCompareOptions, createBeforeScreenshotOptions, buildAfterScreenshotOptions } from './options.js'
import type { ClassOptions } from './options.interfaces.js'
import type { ScreenMethodImageCompareCompareOptions } from '../methods/images.interfaces.js'
import type { InstanceData } from '../methods/instanceData.interfaces.js'
import type { BeforeScreenshotResult } from './beforeScreenshot.interfaces.js'

describe('options', () => {
    describe('defaultOptions', () => {
        it('should return the default options when no options are provided', () => {
            expect(defaultOptions({})).toMatchSnapshot()
        })

        it('should return the provided options when options are provided', () => {
            const options: ClassOptions = {
                addressBarShadowPadding: 1,
                autoSaveBaseline: true,
                formatImageName: '{foo}-{bar}',
                savePerInstance: true,
                toolBarShadowPadding: 1,
                disableBlinkingCursor: true,
                disableCSSAnimation: true,
                fullPageScrollTimeout: 12345,
                hideScrollBars: true,
                blockOutSideBar: true,
                blockOutStatusBar: true,
                blockOutToolBar: true,
                createJsonReportFiles: true,
                diffPixelBoundingBoxProximity: 123,
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
                tabbableOptions: {
                    circle: {
                        backgroundColor: 'backgroundColor',
                        borderColor: 'borderColor',
                        borderWidth: 123,
                        fontColor: 'fontColor',
                        fontFamily: 'fontFamily',
                        fontSize: 321,
                        size: 567,
                        showNumber: false,
                    },
                    line: {
                        color: 'color',
                        width: 987,
                    },
                },
            }

            expect(defaultOptions(options)).toMatchSnapshot()
        })
    })

    describe('methodCompareOptions', () => {
        it('should not return the method options when no options are provided', () => {
            expect(methodCompareOptions({})).toMatchSnapshot()
        })

        it('should return the provided options when options are provided', () => {
            const options = {
                blockOut: [{ height: 1, width: 2, x: 3, y: 4 }],
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
            }

            expect(methodCompareOptions(options)).toMatchSnapshot()
        })
    })

    describe('screenMethodCompareOptions', () => {
        it('should not return the screen method options when no options are provided', () => {
            expect(screenMethodCompareOptions({})).toMatchSnapshot()
        })

        it('should return the provided options when options are provided', () => {
            const options: ScreenMethodImageCompareCompareOptions = {
                blockOutSideBar: false,
                blockOutStatusBar: false,
                blockOutToolBar: false,
                blockOut: [{ height: 1, width: 2, x: 3, y: 4 }],
                ignoreAlpha: true,
                ignoreAntialiasing: true,
                ignoreColors: true,
                ignoreLess: true,
                ignoreNothing: true,
                rawMisMatchPercentage: true,
                returnAllCompareData: true,
                saveAboveTolerance: 12,
                scaleImagesToSameSize: true,
            }

            expect(screenMethodCompareOptions(options)).toMatchSnapshot()
        })
    })

    describe('createBeforeScreenshotOptions', () => {
        const mockElement = { tagName: 'DIV' } as HTMLElement
        const baseWicOptions = {
            addressBarShadowPadding: 10,
            toolBarShadowPadding: 20,
        }

        it('should create options with defaults when minimal options are provided', () => {
            const result = createBeforeScreenshotOptions('testInstance', {}, baseWicOptions)

            expect(result).toEqual({
                instanceData: 'testInstance',
                addressBarShadowPadding: 10,
                toolBarShadowPadding: 20,
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                hideElements: [],
                noScrollBars: false,
                removeElements: [],
                waitForFontsLoaded: false,
            })
        })

        it('should prioritize methodOptions over wicOptions for overlapping properties', () => {
            const methodOptions = {
                disableBlinkingCursor: true,
                disableCSSAnimation: true,
                hideScrollBars: true,
            }
            const wicOptions = {
                ...baseWicOptions,
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                hideScrollBars: false,
            }
            const result = createBeforeScreenshotOptions('testInstance', methodOptions, wicOptions)

            expect(result.disableBlinkingCursor).toBe(true)
            expect(result.disableCSSAnimation).toBe(true)
            expect(result.noScrollBars).toBe(true)
        })

        it('should use wicOptions when methodOptions are undefined', () => {
            const wicOptions = {
                ...baseWicOptions,
                disableBlinkingCursor: true,
                enableLayoutTesting: true,
                waitForFontsLoaded: true,
            }
            const result = createBeforeScreenshotOptions('testInstance', {}, wicOptions)

            expect(result.disableBlinkingCursor).toBe(true)
            expect(result.enableLayoutTesting).toBe(true)
            expect(result.waitForFontsLoaded).toBe(true)
        })

        it('should handle element arrays from methodOptions', () => {
            const hideElements = [mockElement]
            const removeElements = [mockElement, mockElement]
            const methodOptions = {
                hideElements,
                removeElements,
            }
            const result = createBeforeScreenshotOptions('testInstance', methodOptions, baseWicOptions)

            expect(result.hideElements).toBe(hideElements)
            expect(result.removeElements).toBe(removeElements)
        })

        it('should handle all boolean options set to true in methodOptions', () => {
            const methodOptions = {
                disableBlinkingCursor: true,
                disableCSSAnimation: true,
                enableLayoutTesting: true,
                hideScrollBars: true,
                waitForFontsLoaded: true,
            }
            const result = createBeforeScreenshotOptions('testInstance', methodOptions, baseWicOptions)

            expect(result.disableBlinkingCursor).toBe(true)
            expect(result.disableCSSAnimation).toBe(true)
            expect(result.enableLayoutTesting).toBe(true)
            expect(result.noScrollBars).toBe(true)
            expect(result.waitForFontsLoaded).toBe(true)
        })

        it('should preserve instanceData exactly as passed', () => {
            const complexInstanceData = { browser: 'chrome', version: '120', viewport: { width: 1920, height: 1080 } }
            const result = createBeforeScreenshotOptions(complexInstanceData, {}, baseWicOptions)

            expect(result.instanceData).toBe(complexInstanceData)
        })
    })

    describe('buildAfterScreenshotOptions', () => {
        const mockInstanceData: InstanceData = {
            appName: 'testApp',
            browserName: 'chrome',
            browserVersion: '120.0.0',
            deviceName: 'desktop',
            devicePixelRatio: 2,
            deviceRectangles: {
                bottomBar: { height: 0, width: 0, x: 0, y: 0 },
                homeBar: { height: 0, width: 0, x: 0, y: 0 },
                leftSidePadding: { height: 0, width: 0, x: 0, y: 0 },
                rightSidePadding: { height: 0, width: 0, x: 0, y: 0 },
                screenSize: { height: 1080, width: 1920 },
                statusBar: { height: 0, width: 0, x: 0, y: 0 },
                statusBarAndAddressBar: { height: 0, width: 0, x: 0, y: 0 },
                viewport: { height: 900, width: 1200, x: 0, y: 0 }
            },
            initialDevicePixelRatio: 2,
            isAndroid: false,
            isIOS: false,
            isMobile: false,
            logName: 'chrome',
            name: 'chrome',
            nativeWebScreenshot: false,
            platformName: 'desktop',
            platformVersion: '120.0.0'
        }
        const mockEnrichedInstanceData: BeforeScreenshotResult = {
            ...mockInstanceData,
            dimensions: {
                body: {
                    scrollHeight: 1000,
                    offsetHeight: 1000
                },
                html: {
                    clientWidth: 1200,
                    scrollWidth: 1200,
                    clientHeight: 900,
                    scrollHeight: 1000,
                    offsetHeight: 1000
                },
                window: {
                    devicePixelRatio: 3,
                    innerHeight: 800,
                    innerWidth: 1100,
                    isEmulated: false,
                    isLandscape: true,
                    outerHeight: 1000,
                    outerWidth: 1200,
                    screenHeight: 1440,
                    screenWidth: 2560,
                }
            },
            isAndroidChromeDriverScreenshot: false,
            isAndroidNativeWebScreenshot: false,
            isTestInBrowser: true,
            isTestInMobileBrowser: false,
            addressBarShadowPadding: 10,
            toolBarShadowPadding: 20
        }
        const baseInput = {
            base64Image: 'test-screenshot-data',
            folders: { actualFolder: '/test/actual' },
            tag: 'test-element',
            isNativeContext: false,
            instanceData: mockInstanceData,
            wicOptions: {
                formatImageName: '{tag}-{browserName}-{width}x{height}',
                savePerInstance: false
            }
        }

        it('should build options for native commands (no enriched data)', () => {
            const input = {
                ...baseInput,
                isNativeContext: true
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result).toMatchSnapshot()
            expect(result.isNativeContext).toBe(true)
            expect(result.isLandscape).toBe(false)
            expect(result.disableBlinkingCursor).toBeUndefined()
            expect(result.hideElements).toBeUndefined()
        })

        it('should build options for web commands with enriched data', () => {
            const beforeOptions = {
                instanceData: mockInstanceData,
                addressBarShadowPadding: 10,
                toolBarShadowPadding: 20,
                disableBlinkingCursor: true,
                disableCSSAnimation: false,
                enableLayoutTesting: true,
                hideElements: [] as HTMLElement[],
                noScrollBars: true,
                removeElements: [] as HTMLElement[],
                waitForFontsLoaded: false
            }
            const input = {
                ...baseInput,
                enrichedInstanceData: mockEnrichedInstanceData,
                beforeOptions
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result).toMatchSnapshot()
            expect(result.isNativeContext).toBe(false)
            expect(result.isLandscape).toBe(true)
            expect(result.disableBlinkingCursor).toBe(true)
            expect(result.disableCSSAnimation).toBe(false)
            expect(result.enableLayoutTesting).toBe(true)
            expect(result.hideScrollBars).toBe(true)
            expect(result.hideElements).toEqual([])
            expect(result.removeElements).toEqual([])
        })

        it('should prioritize enriched data over instance data', () => {
            const input = {
                ...baseInput,
                enrichedInstanceData: mockEnrichedInstanceData
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.fileName.devicePixelRatio).toBe(3)
            expect(result.fileName.outerHeight).toBe(1000)
            expect(result.fileName.screenHeight).toBe(1440)
            expect(result.isLandscape).toBe(true)
        })

        it('should handle NaN values correctly', () => {
            const enrichedWithNaN: BeforeScreenshotResult = {
                ...mockEnrichedInstanceData,
                dimensions: {
                    ...mockEnrichedInstanceData.dimensions,
                    window: {
                        ...mockEnrichedInstanceData.dimensions.window,
                        devicePixelRatio: undefined,
                        outerHeight: undefined,
                        outerWidth: undefined,
                        screenHeight: undefined,
                        screenWidth: undefined,
                    }
                }
            }
            const input = {
                ...baseInput,
                enrichedInstanceData: enrichedWithNaN
            }

            const result = buildAfterScreenshotOptions(input)

            expect(result.fileName.devicePixelRatio).toBe(2)
            expect(result.fileName.outerHeight).toBeNaN()
            expect(result.fileName.outerWidth).toBeNaN()
            expect(result.fileName.screenHeight).toBe(1080)
            expect(result.fileName.screenWidth).toBe(1920)
        })

        it('should handle missing enriched data gracefully', () => {
            const input = {
                ...baseInput,
                enrichedInstanceData: undefined
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.fileName.devicePixelRatio).toBe(2)
            expect(result.fileName.screenHeight).toBe(1080)
            expect(result.fileName.screenWidth).toBe(1920)
            expect(result.fileName.logName).toBe('chrome')
            expect(result.fileName.name).toBe('chrome')
            expect(result.isLandscape).toBe(false)
        })

        it('should handle element arrays from beforeOptions', () => {
            const mockElement = { tagName: 'DIV' } as HTMLElement
            const beforeOptions = {
                instanceData: mockInstanceData,
                addressBarShadowPadding: 10,
                toolBarShadowPadding: 20,
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                hideElements: [mockElement],
                noScrollBars: false,
                removeElements: [mockElement, mockElement],
                waitForFontsLoaded: false
            }
            const input = {
                ...baseInput,
                beforeOptions
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.hideElements).toEqual([mockElement])
            expect(result.removeElements).toEqual([mockElement, mockElement])
        })

        it('should build correct filePath structure', () => {
            const input = {
                ...baseInput,
                wicOptions: {
                    formatImageName: '{tag}-{logName}',
                    savePerInstance: true
                }
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.filePath).toEqual({
                browserName: 'chrome',
                deviceName: 'desktop',
                isMobile: false,
                savePerInstance: true
            })
        })

        it('should build correct fileName structure with all properties', () => {
            const input = {
                ...baseInput,
                enrichedInstanceData: mockEnrichedInstanceData
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.fileName).toEqual({
                browserName: 'chrome',
                browserVersion: '120.0.0',
                deviceName: 'desktop',
                devicePixelRatio: 3,
                formatImageName: '{tag}-{browserName}-{width}x{height}',
                isMobile: false,
                isTestInBrowser: true,
                logName: 'chrome',
                name: 'chrome',
                outerHeight: 1000,
                outerWidth: 1200,
                platformName: 'desktop',
                platformVersion: '120.0.0',
                screenHeight: 1440,
                screenWidth: 2560,
                tag: 'test-element'
            })
        })

        it('should handle mobile device correctly', () => {
            const mobileInstanceData: InstanceData = {
                ...mockInstanceData,
                isMobile: true,
                deviceName: 'iPhone',
                platformName: 'iOS'
            }
            const mobileEnrichedData: BeforeScreenshotResult = {
                ...mockEnrichedInstanceData,
                isMobile: true,
                deviceName: 'iPhone',
                platformName: 'iOS'
            }
            const input = {
                ...baseInput,
                instanceData: mobileInstanceData,
                enrichedInstanceData: mobileEnrichedData
            }
            const result = buildAfterScreenshotOptions(input)

            expect(result.filePath.isMobile).toBe(true)
            expect(result.fileName.isMobile).toBe(true)
            expect(result.fileName.deviceName).toBe('iPhone')
            expect(result.platformName).toBe('iOS')
        })
    })
})
