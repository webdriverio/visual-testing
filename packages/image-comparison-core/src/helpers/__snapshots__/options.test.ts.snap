// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options > buildAfterScreenshotOptions > should build options for native commands (no enriched data) 1`] = `
{
  "actualFolder": "/test/actual",
  "base64Image": "test-screenshot-data",
  "fileName": {
    "browserName": "chrome",
    "browserVersion": "120.0.0",
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "formatImageName": "{tag}-{browserName}-{width}x{height}",
    "isMobile": false,
    "isTestInBrowser": false,
    "logName": "chrome",
    "name": "chrome",
    "outerHeight": NaN,
    "outerWidth": NaN,
    "platformName": "desktop",
    "platformVersion": "120.0.0",
    "screenHeight": 1080,
    "screenWidth": 1920,
    "tag": "test-element",
  },
  "filePath": {
    "browserName": "chrome",
    "deviceName": "desktop",
    "isMobile": false,
    "savePerInstance": false,
  },
  "isLandscape": false,
  "isNativeContext": true,
  "platformName": "desktop",
}
`;

exports[`options > buildAfterScreenshotOptions > should build options for web commands with enriched data 1`] = `
{
  "actualFolder": "/test/actual",
  "base64Image": "test-screenshot-data",
  "disableBlinkingCursor": true,
  "disableCSSAnimation": false,
  "enableLayoutTesting": true,
  "fileName": {
    "browserName": "chrome",
    "browserVersion": "120.0.0",
    "deviceName": "desktop",
    "devicePixelRatio": 3,
    "formatImageName": "{tag}-{browserName}-{width}x{height}",
    "isMobile": false,
    "isTestInBrowser": true,
    "logName": "chrome",
    "name": "chrome",
    "outerHeight": 1000,
    "outerWidth": 1200,
    "platformName": "desktop",
    "platformVersion": "120.0.0",
    "screenHeight": 1440,
    "screenWidth": 2560,
    "tag": "test-element",
  },
  "filePath": {
    "browserName": "chrome",
    "deviceName": "desktop",
    "isMobile": false,
    "savePerInstance": false,
  },
  "hideElements": [],
  "hideScrollBars": true,
  "isLandscape": true,
  "isNativeContext": false,
  "platformName": "desktop",
  "removeElements": [],
}
`;

exports[`options > defaultOptions > should return the default options when no options are provided 1`] = `
{
  "addIOSBezelCorners": false,
  "addressBarShadowPadding": 6,
  "autoElementScroll": true,
  "autoSaveBaseline": true,
  "clearFolder": false,
  "compareOptions": {
    "blockOutSideBar": true,
    "blockOutStatusBar": true,
    "blockOutToolBar": true,
    "createJsonReportFiles": false,
    "diffPixelBoundingBoxProximity": 5,
    "ignoreAlpha": false,
    "ignoreAntialiasing": false,
    "ignoreColors": false,
    "ignoreLess": false,
    "ignoreNothing": false,
    "rawMisMatchPercentage": false,
    "returnAllCompareData": false,
    "saveAboveTolerance": 0,
    "scaleImagesToSameSize": false,
  },
  "disableBlinkingCursor": false,
  "disableCSSAnimation": false,
  "enableLayoutTesting": false,
  "enableLegacyScreenshotMethod": false,
  "formatImageName": "{tag}-{browserName}-{width}x{height}-dpr-{dpr}",
  "fullPageScrollTimeout": 1500,
  "hideScrollBars": true,
  "isHybridApp": false,
  "savePerInstance": false,
  "tabbableOptions": {
    "circle": {
      "backgroundColor": "#ff0000",
      "borderColor": "#000",
      "borderWidth": 1,
      "fontColor": "#fff",
      "fontFamily": "Arial",
      "fontSize": 10,
      "showNumber": true,
      "size": 10,
    },
    "line": {
      "color": "#000",
      "width": 1,
    },
  },
  "toolBarShadowPadding": 6,
  "userBasedFullPageScreenshot": false,
  "waitForFontsLoaded": true,
}
`;

exports[`options > defaultOptions > should return the provided options when options are provided 1`] = `
{
  "addIOSBezelCorners": false,
  "addressBarShadowPadding": 1,
  "autoElementScroll": true,
  "autoSaveBaseline": true,
  "clearFolder": false,
  "compareOptions": {
    "blockOutSideBar": true,
    "blockOutStatusBar": true,
    "blockOutToolBar": true,
    "createJsonReportFiles": true,
    "diffPixelBoundingBoxProximity": 123,
    "ignoreAlpha": true,
    "ignoreAntialiasing": true,
    "ignoreColors": true,
    "ignoreLess": true,
    "ignoreNothing": true,
    "rawMisMatchPercentage": true,
    "returnAllCompareData": true,
    "saveAboveTolerance": 12,
    "scaleImagesToSameSize": true,
  },
  "disableBlinkingCursor": true,
  "disableCSSAnimation": true,
  "enableLayoutTesting": false,
  "enableLegacyScreenshotMethod": false,
  "formatImageName": "{foo}-{bar}",
  "fullPageScrollTimeout": 12345,
  "hideScrollBars": true,
  "isHybridApp": false,
  "savePerInstance": true,
  "tabbableOptions": {
    "circle": {
      "backgroundColor": "backgroundColor",
      "borderColor": "borderColor",
      "borderWidth": 123,
      "fontColor": "fontColor",
      "fontFamily": "fontFamily",
      "fontSize": 321,
      "showNumber": false,
      "size": 567,
    },
    "line": {
      "color": "color",
      "width": 987,
    },
  },
  "toolBarShadowPadding": 1,
  "userBasedFullPageScreenshot": false,
  "waitForFontsLoaded": true,
}
`;

exports[`options > methodCompareOptions > should not return the method options when no options are provided 1`] = `{}`;

exports[`options > methodCompareOptions > should return the provided options when options are provided 1`] = `
{
  "blockOut": [
    {
      "height": 1,
      "width": 2,
      "x": 3,
      "y": 4,
    },
  ],
  "ignoreAlpha": true,
  "ignoreAntialiasing": true,
  "ignoreColors": true,
  "ignoreLess": true,
  "ignoreNothing": true,
  "rawMisMatchPercentage": true,
  "returnAllCompareData": true,
  "saveAboveTolerance": 12,
  "scaleImagesToSameSize": true,
}
`;

exports[`options > screenMethodCompareOptions > should not return the screen method options when no options are provided 1`] = `{}`;

exports[`options > screenMethodCompareOptions > should return the provided options when options are provided 1`] = `
{
  "blockOut": [
    {
      "height": 1,
      "width": 2,
      "x": 3,
      "y": 4,
    },
  ],
  "blockOutSideBar": false,
  "blockOutStatusBar": false,
  "blockOutToolBar": false,
  "ignoreAlpha": true,
  "ignoreAntialiasing": true,
  "ignoreColors": true,
  "ignoreLess": true,
  "ignoreNothing": true,
  "rawMisMatchPercentage": true,
  "returnAllCompareData": true,
  "saveAboveTolerance": 12,
  "scaleImagesToSameSize": true,
}
`;
